kolibri/auth/backends.py

```diff
@@ -146,3 +152,79 @@ def get_all_permissions(self, user_obj, obj=None):
         :return: A list of permission strings. Empty if the user is a FacilityUser.
         """
         raise NotImplementedError()
+
+
+class InvalidPermission(KolibriError):
+    pass
+
+
+def _reject_obj(perm, obj):
+    if obj is not None:
+        raise InvalidPermission("'{perm}' does not take an optional object. Got: {obj}".format(
+            perm=perm, obj=repr(obj)))
+
+
+def _deny(perm, user, obj=None):
+    _reject_obj(perm, obj)
+    return False
+
+
+def _admin_only(perm, user, obj):
+    _reject_obj(perm, obj)
+    return user.is_facility_admin()
+
+
+def _assert_type(obj, _type):
+    if isinstance(obj, _type):
+        return True
+    else:
+        raise InvalidPermission('Expected object of type {}, but got {}'.format(repr(_type), repr(type(obj))))
+
+
+def _coach_for_the_class(user, obj):
+    """
+    Permission formula for auth.change_classroom and auth.remove_classroom
+
+    :param user: A FacilityUser object
+    :param obj: The optional permissions object. Raises an InvalidPermission error if obj is not a Classroom.
+    :return: True if the user is a Coach for the Classroom obj & True if the user is a FacilityAdmin, otherwise False
+    """
+    if obj is not None and _assert_type(obj, Classroom):
+        return user.is_facility_admin() or (user in [role.user for role in obj.coaches()])
+    else:
+        return user.is_facility_admin()
+
+
+def _coach_for_the_learner_group(user, obj):
+    """
+    Permission formula for auth.remove_learner_group and auth.change_learner_group
+
+    :param user: A FacilityUser object
+    :param obj: The optional permissions object. Raises an InvalidPermission error if obj is not a LearnerGroup.
+    :return: True if the user is a Coach for the LearnerGroup's Classroom obj
+             & True if the user is a FacilityAdmin, otherwise False
+    """
+    if obj is not None and _assert_type(obj, LearnerGroup):
+        classroom = obj.classroom()
+        return user.is_facility_admin() or (user in [role.user for role in classroom.coaches()])
+    else:
+        return user.is_facility_admin()
+
+
+_permissions_checkers = {
```

Comment:
isn't it better to add these as decorators on the functions themselves so it's explicit when you read the function def?